src/src_main/crypto_tc.c

```diff
@@ -428,31 +467,56 @@ int32_t Crypto_TC_ApplySecurity(const uint8_t* p_in_frame, const uint16_t in_fra
         */
         // TODO: Revisit this
         // TODO: Likely SA API Call
-        for (i = 0; i < sa_ptr->shplf_len; i++)
+        /* 4.1.1.5.2 The Pad Length field shall contain the count of fill bytes used in the
+        ** cryptographic process, consisting of an integral number of octets. - CCSDS 3550b1
+        */
+        // TODO: Set this depending on crypto cipher used
+
+        if(padding_needed)
         {
-            /* 4.1.1.5.2 The Pad Length field shall contain the count of fill bytes used in the
-            ** cryptographic process, consisting of an integral number of octets. - CCSDS 3550b1
-            */
-            // TODO: Set this depending on crypto cipher used
-            *(p_new_enc_frame + index) = 0x00;
-            index++;
+            char hex_padding[3] = {0};  //TODO: Create #Define for the 3
+            padding_needed = padding_needed & 0x00FFFFFF; // Truncate to be maxiumum of 3 bytes in size
+            
+            // Byte Magic
+            hex_padding[0] = (padding_needed >> 16) & 0xFF;
+            hex_padding[1] = (padding_needed >> 8)  & 0xFF;
+            hex_padding[2] = (padding_needed)  & 0xFF;
+            
+
+            uint8_t padding_start = 0;
+            padding_start = 3 - sa_ptr->shplf_len;
+
+            for (i = 0; i < sa_ptr->shplf_len; i++)
+            {
+                *(p_new_enc_frame + index)  = hex_padding[padding_start++];
+                index++;
+            }
         }
 
+        
         /*
         ** End Security Header Fields
         */
 
-        // Determine if FECF exists
-        uint8_t fecf_len = FECF_SIZE;
-        if (current_managed_parameters->has_fecf == TC_NO_FECF)
-        {
-            fecf_len = 0;
-        }
+        
 
         // Copy in original TF data - except FECF
         // Will be over-written if using encryption later
-        tf_payload_len = temp_tc_header.fl - TC_FRAME_HEADER_SIZE - segment_hdr_len - fecf_len + 1;
+        //tf_payload_len = temp_tc_header.fl - TC_FRAME_HEADER_SIZE - segment_hdr_len - fecf_len + 1;
+        
         memcpy((p_new_enc_frame + index), (p_in_frame + TC_FRAME_HEADER_SIZE + segment_hdr_len), tf_payload_len);
+        index += tf_payload_len;
+        for (uint32_t i = 0; i < padding_needed; i++)
+        {
+            /* 4.1.1.5.2 The Pad Length field shall contain the count of fill bytes used in the
+            ** cryptographic process, consisting of an integral number of octets. - CCSDS 3550b1
+            */
+            // TODO: Set this depending on crypto cipher used
+            *(p_new_enc_frame + index + i) = (uint8_t)padding_needed; // How much padding is needed?
```

Comment:
Should this be equal to "padding_needed" or just a padding byte? EG 0xFF?

Reply:
It needs to be equal to padding_needed.  Each byte of padding needs to be specific.  EG:
1 byte of padding:  padding = 0x01
2 bytes: 0202
3 : 030303 etc.

Reply:
Very interesting -- So padding_needed is actually the 4 bytes of padding that are needed -- What reference explains how the bytes are padding are determined?

Reply:
Nevermind, I see the PKCS#7 padding reference in wiki: https://en.wikipedia.org/wiki/Padding_(cryptography)#PKCS7

Reply:
This is based on the PKCS7/PKCS5 standards.
If your block size is 16, and you have 14 bytes of padding, you need 2 bytes of padding 0202.
If you have 16 bytes, you actually need another 16 bytes of padding 101010101010...

> PKCS5 Padding
If the block length is B then add N padding bytes of value N to make the input length up to the next exact multiple of B. If the input length is already an exact multiple of B then add B bytes of value B. Thus padding of length N between one and B bytes is always added in an unambiguous manner. After decrypting, check that the last N bytes of the decrypted data all have value N with 1 < N ≤ B. If so, strip N bytes, otherwise throw a decryption error.

Examples of PKCS5 padding for block length B = 8:

3 bytes: FDFDFD           --> FDFDFD0505050505
7 bytes: FDFDFDFDFDFDFD   --> FDFDFDFDFDFDFD01
8 bytes: FDFDFDFDFDFDFDFD --> FDFDFDFDFDFDFDFD0808080808080808

Reply:
Are you sure you're supposed to pad when the block size is equal to block length? I'll have to look closer at your code to see if that actually happens. I wouldn't think so 🤔 ...

Reply:
Yup.  I'm sure.  KMC does it as well.  There are unit tests verifying it.  Feel free to research, though.